### PR TITLE
Total Difficulty in IBFT

### PR DESCRIPTION
**Pull request by dbrajovic**
# Description

Defining total difficulty in our IBFT implementation.
Addresses [#179](https://github.com/0xPolygon/polygon-sdk/issues/179)

Update:

Hardcoding `GenesisDifficulty` to 1 breaks backwards compatibility for existing chains due to new nodes having to generate the genesis block with difficulty 1.

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [x] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have tested this code
- [ ] I have updated the README and other relevant documents (guides...)
- [ ] I have added sufficient documentation both in code, as well as in the READMEs

# Additional comments

This PR is about coming to terms on what Total Difficulty means in IBFT.

In other consensuses, TD is what helps identify the correct chain (fork) based on the sum of all blocks' `difficulty` field (which varies from block to block). 
In IBFT, TD has nothing to do with the `difficulty` field, but serves the same purpose, in a sense that it decides what is the latest block (number) - hence this value should represent the block's `number` field.

In short (our IBFT impl):
```
Block's difficulty -> Block's number
Block's TD -> Sum of all previous block numbers (up to current)
Chain TD -> TD of the latest block in the chain
```

Related issues:
[https://github.com/0xPolygon/polygon-sdk/issues/179](#179)

This fix should provide the `Syncer` with the correct value (fetched with `CurrentTD()`) in order to resolve its best peer.

### CODE REVIEW

**Comment by dbrajovic** [comment]
After tuning the unit tests, I discovered that TD is calculated in other places as well (`advanceHead()`).  At all points, there is a fetching of the parent difficulty used only to calculate new TD (which is redundant in this new approach). 

I've left `_ = variableName` to suppress compiler warnings - I intend to remove these once proven that they were not meant for anything else.

Regarding the TD field in the units test (blockchain_test.go), this is the value we would like to see hardcoded to 1, in which case, all that is needed is to hardcode `GetChainTD()` to always return 1.

**Comment by dbrajovic** [comment]
After revising the aforementioned calculation of TD, I decided to go with the less invasive approach of hardcoding each block's difficulty to 1, keeping the existing logic (summing of difficulties) intact, such that the following are true:

> Block difficulty = 1
> Block TD = Block number (1 + 1 + ... + 1)
> ChainTD = TD of the latest block (its number)

File: consensus/ibft/ibft.go [hunk 1]
```diff
@@ -880,8 +880,8 @@ func (i *Ibft) verifyHeaderImpl(snap *Snapshot, parent, header *types.Header) er
 		return fmt.Errorf("invalid sha3 uncles")
 	}
 
-	// difficulty has to match number
-	if header.Difficulty != header.Number {
+	// difficulty has to be 1 (except for genesis block)
+	if header.Difficulty != 1 && header.Number != 0 {
```
Comment (zivkovicmilos) [inline consensus/ibft/ibft.go]: Why does the genesis block have a difficulty != 1?

**Review by zivkovicmilos — CHANGES_REQUESTED** [review]
Looks great to me 💯 

Left a minor question that needs clarification

File: consensus/ibft/ibft.go (hunk 1 above)
Comment (dbrajovic) [reply consensus/ibft/ibft.go]: Misplaced difficulty check. The failing `TestTransition_AcceptState_Validator_VerifyCorrect` unit test lead me to that conclusion.

[Fix](https://github.com/0xPolygon/polygon-sdk/blob/a39a21bb547f99417d645b59ecf087fec2835371/chain/chain.go#L20) (breaking change)

**Review by zivkovicmilos — APPROVED** [review]
Thank you for making the requested changes 🙏 

Looks great now 💯

File: blockchain/blockchain.go [hunk 2]
```diff
@@ -285,21 +285,21 @@ func (b *Blockchain) GetTD(hash types.Hash) (*big.Int, bool) {
...
 func (b *Blockchain) writeCanonicalHeader(event *Event, h *types.Header) error {
-	td, ok := b.readDiff(h.ParentHash)
+	parentTD, ok := b.readDiff(h.ParentHash)
 	if !ok {
 		return fmt.Errorf("parent difficulty not found")
 	}
 
-	diff := big.NewInt(1).Add(td, new(big.Int).SetUint64(h.Difficulty))
-	if err := b.db.WriteCanonicalHeader(h, diff); err != nil {
+	currentTD := big.NewInt(1).Add(parentTD, new(big.Int).SetUint64(h.Difficulty))
```
Comment (Kourin1996) [inline blockchain/blockchain.go]: I think this won't cause a misleading. (Just I thought why `1` here?)

```go
currentTD := new(big.Int).Add(parentTD, new(big.Int).SetUint64(h.Difficulty))
```

or

```go
currentTD := big.NewInt(0).Add(parentTD, new(big.Int).SetUint64(h.Difficulty))
```

File: blockchain/blockchain.go [hunk 3]
```diff
@@ -322,25 +322,25 @@ func (b *Blockchain) advanceHead(newHeader *types.Header) (*big.Int, error) {
...
 			return nil, fmt.Errorf("parent difficulty not found")
 		}
-		currentDiff = td
+		parentTD = td
 	}
 
 	// Calculate the new difficulty
-	diff := big.NewInt(1).Add(currentDiff, new(big.Int).SetUint64(newHeader.Difficulty))
-	if err := b.db.WriteDiff(newHeader.Hash, diff); err != nil {
+	newTD := big.NewInt(1).Add(parentTD, new(big.Int).SetUint64(newHeader.Difficulty))
```
Comment (Kourin1996) [inline blockchain/blockchain.go]: This is also

```go
newTD := new(big.Int).Add(parentTD, new(big.Int).SetUint64(newHeader.Difficulty))
```

**Review by Kourin1996 — COMMENTED** [review]
Thank you for the PR. I like renaming variable to proper one. I've left small comments. I'd appreciated if you could check and fix if need.

File: blockchain/blockchain.go (hunk 2 above)
Comment (dbrajovic) [reply blockchain/blockchain.go]: Previous implementation logic. 

Addressed in this [commit](https://github.com/0xPolygon/polygon-sdk/pull/182/commits/8f320b13ff4c911d5c7165d8c2259910ba05c495)

File: blockchain/blockchain.go (hunk 3 above)
Comment (dbrajovic) [reply blockchain/blockchain.go]: [Fixed](https://github.com/0xPolygon/polygon-sdk/pull/182/commits/8f320b13ff4c911d5c7165d8c2259910ba05c495)

**Review by Kourin1996 — APPROVED** [review]
Thank you for fixing. LGTM👍

**Review by zivkovicmilos — APPROVED** [review]
Approving for a second time because Github reset my approval for some reason 🙂

**Comment by lazartravica** [comment]
@dbrajovic feel free to close this PR and delete the branch if you plan on opening a new one 👍

**Comment by lazartravica** [comment]
Closing this PR in favour of https://github.com/0xPolygon/polygon-sdk/pull/208

cc @dbrajovic
